Remora.Rest/Json/Internal/BoundDataObjectConverter.cs

```diff
@@ -0,0 +1,272 @@
+            var write = source.GetPropertyWriter(property);
+
+            // We cache this as well since the check is somewhat complex
+            bool allowsNull = property.AllowsNull();
+
+            var data = new DTOPropertyInfo
+            (
+                property,
+                readNames,
+                writeNames,
+                write,
+                allowsNull,
+                defaultValue,
+                propertyOptions,
+                readProperties.Count
+            );
+
+            if (property.CanWrite)
+            {
+                // If a property is writable, it can be *read* from JSON.
+                readProperties.Add(data);
+            }
+
+            if (property.CanWrite || source.IncludesReadOnlyProperty(property))
+            {
+                // Any property that is writable and not excluded due to being read-only,
+                // can be *written* to JSON.
+                writeProperties.Add(data);
+            }
+        }
+
+        _writeProperties = writeProperties.ToArray();
+        _readProperties = readProperties.ToArray();
+
+        _readPropertiesByName = readProperties
+            .SelectMany(p => p.ReadNames.Select((n, i) => (IsPrimary: i == 0, n, DTOProperty: p)))
+            .ToDictionary(x => x.n, x => (x.IsPrimary, x.DTOProperty));
+    }
+
+    private static JsonSerializerOptions GetPropertyConverterOptions(JsonSerializerOptions options, JsonConverter? converter)
+    {
+        if (converter == null)
+        {
+            return options;
+        }
+
+        var cloned = new JsonSerializerOptions(options);
+        cloned.Converters.Insert(0, converter);
+
+        return cloned;
+    }
+
+    /// <inheritdoc />
+    public override TInterface Read
+    (
+        ref Utf8JsonReader reader,
+        Type typeToConvert,
+        JsonSerializerOptions options
+    )
+    {
+        if (reader.TokenType != JsonTokenType.StartObject ||
+            !reader.Read())
+        {
+            DataObjectConverterShared.ThrowJsonException();
+        }
+
+        var readProperties = _readProperties;
+
+        // CMBK: Considering to cache this array and reuse it.
+        // F.e. in a thread-static field. This method isn't async anyways.
+
+        // Avoid creating a dictionary and fill the arguments array directly
+        var constructorArguments = new object?[readProperties.Length];
+
+        // We use "Missing" as a sentinel to indicate that a value hasn't
+        // been seen yet without needing to allocate additional memory.
+        constructorArguments.AsSpan().Fill(DataObjectConverterShared.Missing);
+
+        while (reader.TokenType != JsonTokenType.EndObject)
+        {
+            var propertyName = reader.GetString();
+            if (!reader.Read())
+            {
+                DataObjectConverterShared.ThrowJsonException();
+            }
+
+            Debug.Assert(propertyName != null, "Expected property names to never be null.");
```

Comment:
Throw exception or remove.

Reply:
STJ would have already thrown in we had null as a property name, however `GetString` returns null. Either it's this, or I need to suppress a warning about null right below it.

Reply:
I'd say the nullability suppression is more appropriate in this case; we know for a fact that the property name is not null at this point. I can also see the argument for `Debug.Assert` being more expressive, too, though.